docs/multi-datasource/high_level_design.md

```diff
@@ -0,0 +1,147 @@
+- data source capabilities, such as if the data source supports AD/ISM etc.
+
+Users can dynamically add data source in OpenSearch Dashboards using UI or API, OpenSearch Dashboards will save the data source saved objects in its metadata index. And then users can do their with with their data sources. For example, when OpenSearch Dashboards needs to access user data on behalf of the customer, customer will need to specify a data source id, then OpenSearch Dashboards can fetch the data source info from its metadata store, then send the request to the corresponding data source endpoint.
+
+So the Dashboards and OpenSearch setup may look like:![img](./img/hld_setup_diagram.png)
+
+Refer to the proposed solution in [#1388](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1388) for the data modeling of data source
+
+### Data source integration
+
+[opensearch_service](https://github.com/opensearch-project/OpenSearch-Dashboards/tree/main/src/core/server/opensearch) is one of the core modules of OpenSearch Dashboards, it is a singleton instance in OpenSearch Dashboards which manages OpenSearch Dashboards connection with the backend OpenSearch endpoint. It makes another level of abstraction of OpenSearch client, and provide a set of interfaces for other OpenSearch Dashboards modules and plugins to interact with OpenSearch for example running DSL queries, or calling arbitrary OpenSearch APIs.
+
+Currently, OpenSearch Dashboards only works with one OpenSearch cluster, OpenSearch Dashboards metadata index and user data indices are stored in the same OpenSearch cluster. So the OpenSearch Dashboards [saved object service](https://github.com/opensearch-project/OpenSearch-Dashboards/tree/main/src/core/server/saved_objects), which the core OpenSearch Dashboards module handles all OpenSearch Dashboards metadata operations, also relies on `opensearch_service` interfaces to work with OpenSearch.
+
+With multi-datasource, we will need to diverge the `opensearch_service` for these 2 use cases. We propose to fork a new `metadata_client` from existing `opensearch_service` to manage the metadata store connection, so that `saved_objects_service` can use `metadata_client` to perform saved objects operations. And then we repurpose the `opensearch_service` to serve the user data access use cases. The new `opensearch_service` needs will expose following interface to allow other OpenSearch Dashboards components to interact with a specific data source cluster.
+
+```
+core.opensearch.withDataSource(<data_source_id>).callAsCurrentUser(searchParams)
+```
+
+OpenSearch Dashboards plugins like data plugin, alerting plugin will need to introduce the data source concept into their use case, letting users to specify a data source when using their functions, and then switch to this new opensearch interface when calling OpenSearch APIs or executing queries.
+
+### Multi-datasource support in visualizations
```

Comment:
This section is a lot more powerful than it appears to be and i think it deserves a bigger callout. The title is a little misleading too 😛. Based on what I've read so far, **any feature** including visualizations and dashboards should work out of the box with no changes since the multi datasource feature is transparent to the features that rely on the higher level search api and services like saved objects. 

This section currently makes it clear that visualizations and dashboards dont need any change, but that callout is at the end and does not paint a full picture of how nice this design is. Even other features (e.g. Vis builder) works out of the box because of how transparent this feature is to other high level apps

Reply:
Like I called out earlier, can we update this section to be a little clearer? This feels like an "Integrating saved objects with Multi datasource" kinda section and not visualizations specifically. Would it be possible to describe when and how to integrate with it? That was my takeaway from this section.

What i'm looking for here is a more general summary of the relationship between saved objects and milti datasource. We have a lot more than 3 saved object types and afaik, only index patterns need to be modified to support muti datasources since they are the object that all other objects reference, and the one that need to know which datasource to call.

Reply:
"data source " itself is also a "saved object". Not every saved object is related to multiple data source feature. For example, ui-settings, dql query, has nothing to do with data source, so "integrating saved objects with multiple datasource" is not the point we try to make here. We are only referencing "data source " to "index-pattern", as a high level mental model to answer the question how to introduce data source to visualization. The referencing itself is a common way how save objects to integrate with each other, besides it's impl detail, and this doc only talks about high level stuff.  

For other saved object type such as vis-builder, if they reference index-pattern, or reference any save object that references index-pattern. Getting data from data source will be supported out of the box (I'll added it to this section)

This section focuses on explaining the change of a common use case after adopting MD feature, which is visualization in OSD. Word visualization here doesn't actually mean the visualization saved object, it's the solution. 

I updated the wording and title a little bit.
@ashwin-pc